src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -469,15 +452,114 @@ static uct_iface_internal_ops_t uct_cuda_copy_iface_internal_ops = {
     .ep_is_connected       = uct_base_ep_is_connected
 };
 
+static ucs_status_t
+uct_cuda_copy_ctx_create_validator(CUcontext ctx,
+                                   uct_cuda_copy_ctx_validator_t *validator)
+{
+#if CUDA_VERSION >= 12000
+    *validator = ctx;
+    return UCS_OK;
+#else
+    return UCT_CUDADRV_FUNC_LOG_ERR(cuMemAlloc(validator, 1));
+#endif
+}
+
+static void
+uct_cuda_copy_ctx_destroy_validator(uct_cuda_copy_ctx_validator_t validator)
+{
+#if CUDA_VERSION < 12000
+    return UCT_CUDADRV_FUNC_LOG_WARN(cuMemFree(validator));
+#endif
+}
+
+static ucs_status_t uct_cuda_copy_ctx_rsc_init(CUcontext ctx,
+                                               unsigned int max_cuda_events,
+                                               uct_cuda_copy_ctx_rsc_t *ctx_rsc)
+{
+    ucs_mpool_params_t mp_params;
+    ucs_status_t status;
+    ucs_memory_type_t src, dst;
+    uct_cuda_copy_iface_mpool_priv_t *mpool_priv;
+
+#if CUDA_VERSION >= 12000
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetId(ctx, &ctx_rsc->ctx_id));
+    if (status != UCS_OK) {
+        return status;
+    }
+#endif
+
+    status = uct_cuda_copy_ctx_create_validator(ctx, &ctx_rsc->validator);
```

Comment:
can you keep validator just in mpool_priv and initialize it after mpool creation?
why is it needed to also keep it in the context?

Reply:
Removed extra field.